SPEC.md

```diff
@@ -0,0 +1,271 @@
+
+The most pervasive pattern in the payments contract is the usage of pre and post condition modifiers. The bulk of these modifier calls force invariants within the fields of the three core datastructures to be true. The major invariant being enforced is that accounts are always settled as far as possible. In fact function modifiers is the only place where account settlement occurs (for more detail see [section below](#accounts-and-account-settlement)). Additionally there are invariants making sure that rails don't attempt to spend more than their fixed lockup and that account locked funds are always covered by account balance. There are also selectively used invariants asserting that rails are in particular termination states for particular methods.
+
+Every interesting function modifying the state of the payments contract runs a group of core account settlement related invariant pre and post conditions via the `settleAccountLockupBeforeAndAfter` or the `settleAccountLockupBeforeAndAfterForRail` modifier. This is a critical mechanism to be aware of when reasoning through which invariants apply during the execution of payments contract methods.
+
+## Operator Approval
+
+As describe above operator approvals consist of the pair of `rateAllowance` and `lockupAllowance`.  Approvals are per operator and rate and lockup resource usage are summed across all of an operator's rails when checking for sufficient operator approval during rail operations.  Approvals also include a `maxLockupPeriod` restricting the operator's ability to make lockup period too long.
+
+The OperatorApproval struct 
+
+```solidity
+    struct OperatorApproval {
+        bool isApproved;
+        uint256 rateAllowance;
+        uint256 lockupAllowance;
+        uint256 rateUsage; // Track actual usage for rate
+        uint256 lockupUsage; // Track actual usage for lockup
+        uint256 maxLockupPeriod; // Maximum lockup period the operator can set for rails created on behalf of the client
+    }
+```
+
+An important counterintuitive fact about the approval allowances is that they are not constrained in relation to current usage. Usage can be lower than allowance if an operator has not used all of their existing allowance. Usage can be higher than allowance if a client has manually reduced the operator's allowance. As explained in the README reducing allowance below usage on any of the allowance resources (rate, lockup, period) will not impact existing rails. Allowance invariants are checked at the point in time of rail modification not continuously enforced. Furthermore reductions in usage always go through even if the current allowance is below the new usage. For example if a rail has an allowance of 20 locked tokens and uses all of them to lock up 20 tokens, and then the client brings allowance for the operator down to 1 locked token the operator can still modify the rail usage down to 15 locked tokens even though it exceeds the operator's current allowance.
```

Comment:
However, new rails created after a "a client has manually reduced the operator's allowance" will have to respect the new allowance limits